Topology widget for containers provider.

abonas opened the pull request:
Displays a graph of container related entities and
their relationships, as well as cross provider relationships
with vms and hosts if such exist.
The patch adds a new UI screen, menu, navigation and backend
logic that enables the topology.

![topologylatest](https://cloud.githubusercontent.com/assets/6277245/8766604/1eebb3ee-2e48-11e5-9932-7e524ad0482f.png)

abonas commented:
@h-kataria , @dclarizio Please review.
I'm wondering if haml conversion is needed here, it seems not trivial with the angular directives and the svg elements.
Also, can you please review the permissions part, it seems that something is missing, because from time to time I get "not authorized" when accessing topology directly, it doesn't reproduce all the time.
thank you.

abonas commented:
@epwinchell the icon will be probably replaced, new design is in process :)

abonas commented:
@miq-bot add_label enhancement
@miq-bot add_label providers/containers

epwinchell commented on app/views/container_topology/show.html:
```diff
@@ -0,0 +1,155 @@
+      margin: 0px;
+    }
+    kubernetes-topology-graph {
+      border: 10px solid lightgray;
```
@abonas I'd suggest changing the border styling to:
border-top: 1px solid #eee;

epwinchell replied:
@abonas I don't see the style change to the border. The fat border around the view is not in keeping with Paternfly or the rest of the UI. Can you update the screenshot once the change has been made?

abonas replied:
> @abonas I don't see the style change to the border. The fat border around the view is not in keeping with Paternfly or the rest of the UI. Can you update the screenshot once the change has been made?

yes I will change it, and update.

epwinchell commented on app/views/container_topology/show.html:
```diff
@@ -0,0 +1,155 @@
+      margin-top: 15px;
+    }
+
+    kubernetes-topology-icon {
```
@abonas Can you remove the kubernetes prefix from these classes?

abonas replied:
> @abonas Can you remove the kubernetes prefix from these classes?

they are named like that in the third party component which is used here
https://github.com/kubernetes-ui/topology-graph/blob/master/topology-graph.css

epwinchell replied:
@abonas I'm sorry to hear this hasn't been generalize. @fryguy @dclarizio Can we use this gem in other parts of the UI or does the project have to be forked and modified?

abonas replied:
> @abonas I'm sorry to hear this hasn't been generalize. @fryguy @dclarizio Can we use this gem in other parts of the UI or does the project have to be forked and modified?

@epwinchell can you please explain what is the problem with the name?
Forking is rarely a good idea and will not hold for a long term due to the constant need of resync and maintenance.
Adding @stefwalter who is the author of the topology component.

Fryguy replied:
There was a thought that this widget was generic and could be used for, say, vmware topology.  In its present form it can't.  I think the question is whether or not this has to be forked in order to make it more generic.

abonas replied:
> There was a thought that this widget was generic and could be used for, say, vmware topology. In its present form it can't. I think the question is whether or not this has to be forked in order to make it more generic.

@Fryguy
OK now I see what you meant by addressing the naming.
The initial requirement for this was to be for containers provider, and the component is currently used in 2 other projects focused on containers management areas.
Lets see what @stefwalter thinks on generalizing.
I'd really recommend to avoid forking. Who is supposed to maintain the sync of the fork with the original component?

cc @simon3z @itamarh

abonas replied:
> There was a thought that this widget was generic and could be used for, say, vmware topology. In its present form it can't. I think the question is whether or not this has to be forked in order to make it more generic.

@Fryguy
OK now I see what you meant by addressing the naming.
The initial requirement for this was to be for containers provider, and the component is currently used in 2 other projects focused on containers management areas.
Lets see what @stefwalter thinks on generalizing.
I'd really recommend to avoid forking. Who is supposed to maintain the sync of the fork with the original component?

cc @simon3z @itamarh

abonas commented:
@h-kataria please review. thanks

dclarizio commented:
@epwinchell please review for style/consistency with existing UI.  Thx, Dan

itamarh commented:
I really hope the widget can be generalized. I'd expect the legend/list of entities is not the issue and can be dynamic, rather the format on actual entities is currently "kubernetes-like". any reason we can't provide the other entities in a similar format for the widget?

Fryguy commented:
@abonas Maybe fork is the wrong word...perhaps "refactoring to a common core"?  I think the thought is that the topology widget is generic, and you initialize it with a list of node types and their icons.  Then you give it the raw data so it can draw the chart.  Right now that list of node types is hardcoded into the widget (from what I can tell), right up to the style names and icons.

Fryguy commented on app/controllers/container_topology_controller.rb:
```diff
@@ -0,0 +1,105 @@
+
+    nodes.each do |n|
+      topo_items[n.ems_ref] = {:metadata => {:id  => n.ems_ref, :name => n.name}, :kind => "Node"}
+      if n.container_groups.size > 0
```
We should use Rubyisms here like `if n.container_groups.any?`

Actually...on second review this line isn't needed at all...when you have no items, .each is a noop, so let's just remove this.

abonas replied:
> We should use Rubyisms here like if n.container_groups.any?
> 
> Actually...on second review this line isn't needed at all...when you have no items, .each is a noop, so let's just remove this.

will do

Fryguy commented on app/controllers/container_topology_controller.rb:
```diff
@@ -0,0 +1,105 @@
+        end
+
+      end
+      unless n.lives_on.nil?
```
Try to avoid double negative cause it's confusing...here you can just say, `if n.lives_on`

abonas replied:
> Try to avoid double negative cause it's confusing...here you can just say, if n.lives_on

good idea

Fryguy commented on app/controllers/container_topology_controller.rb:
```diff
@@ -0,0 +1,105 @@
+
+      end
+      unless n.lives_on.nil?
+        kind = n.lives_on.type.end_with?('Vm') ? "VM" : "Host"
```
This might be bad...avoid String manipulation especially on the Type column.  Prefer `n.lives_on.kind_of?(Vm) ? "Vm" : "Host"`.  @matthewd's rewrite will completely screw up this code.

Or perhaps just use the dictionary or I18N stuff to convert the object you have into a String (i.e. `ui_lookup(:model => n.lives_on.class)`, or something like that...@h-kataria might be able to explain that better.)

abonas replied:
> This might be bad...avoid String manipulation especially on the Type column. Prefer n.lives_on.kind_of?(Vm) ? "Vm" : "Host". @matthewd's rewrite will completely screw up this code.
> 
> Or perhaps just use the dictionary or I18N stuff to convert the object you have into a String (i.e. ui_lookup(:model => n.lives_on.class), or something like that...@h-kataria might be able to explain that better.)

@Fryguy @matthewd 
The below returns false so it seems like it doesn't work with all types of Vms after refactor:
VmOpenstack.kind_of?(Vm)

chessbyte replied:
VmOpenstack is a class.  `VmOpenstack.new.kind_of?(Vm)` returns true.

abonas replied:
> @abonas VmOpenstack is a class. VmOpenstack.new.kind_of?(Vm) returns true.
> cc @Fryguy @matthewd

good catch, thanks

Fryguy commented on app/controllers/container_topology_controller.rb:
```diff
@@ -0,0 +1,105 @@
+        if kind == 'VM' # add link to Host
+          host = n.lives_on.host
+          topo_items[host.uid_ems] = {:metadata => {:id => host.uid_ems, :name => host.name}, :kind => 'Host'}
+          links << {:source => n.lives_on.uid_ems, :target => host.uid_ems}
```
You already have a build_link method...why isn't it being used?

Fryguy commented on app/controllers/container_topology_controller.rb:
```diff
@@ -0,0 +1,105 @@
+        topo_items[n.lives_on.uid_ems] = {:metadata => {:id   => n.lives_on.uid_ems,
+                                                        :name => n.lives_on.name},
+                                          :kind     => kind}
+        links << {:source => n.ems_ref, :target => n.lives_on.uid_ems}
```
You already have a build_link method...why isn't it being used?

Fryguy commented on app/controllers/container_topology_controller.rb:
```diff
@@ -0,0 +1,105 @@
+        links << {:source => n.ems_ref, :target => n.lives_on.uid_ems}
+        if kind == 'VM' # add link to Host
+          host = n.lives_on.host
+          topo_items[host.uid_ems] = {:metadata => {:id => host.uid_ems, :name => host.name}, :kind => 'Host'}
```
And here you can use the build_entity method.

abonas replied:
> And here you can use the build_entity method.

I refactored the whole build_X methods thing to be able to use them everywhere through this code

Fryguy commented on app/helpers/application_helper/page_layouts.rb:
```diff
@@ -2,9 +2,11 @@ module ApplicationHelper::PageLayouts
 
   def layout_uses_listnav?
     if !(@layout == "dashboard" && ["show","change_tab","auth_error"].include?(controller.action_name) ||
+         @layout.nil? ||
```
Why the nil check?

abonas replied:
> Why the nil check?

good catch, it was here from topology first days. removing it...

Fryguy commented on app/views/container_topology/show.html.haml:
```diff
@@ -0,0 +1,122 @@
+    width: 39px !important;;
+    height: 39px !important;;
+    display: block;
+  }
```
Should css really be embedded in the haml?  @epwinchell Thoughts?

himdel replied:
Can't speak for css in general, but the first two rules should be unnecessary since bootstrap already does css reset.

epwinchell replied:
The css should move into a topology.css file in assets/stylesheets

abonas replied:
> Can't speak for css in general, but the first two rules should be unnecessary since bootstrap already does css reset.

If you are talking about the "*" and the body rules, removing them causes the topology legend labels to have a different font than the rest of ManageIQ UI a 

> Can't speak for css in general, but the first two rules should be unnecessary since bootstrap already does css reset.

removing first 2 rules (I assume you talk about "*" and "body") causes the page to look differently, for example the text in labels in legend has a different font than the rest of the text in ManageIQ menus. (just to clarify, the topology widget external component is also bringing its own css, and I am overriding here whatever should be overriden for functional and visual reasons)

himdel replied:
@abonas The point is, there is no scoping style, you're overriding CSS for the whole application, not just your part.

abonas replied:
> The css should move into a topology.css file in assets/stylesheets

what is the correct way to reference a css in a haml file when everything is moved to a separate css file? I saw in one of the existing haml this "stylesheet_link_tag" but it doesn't refer to any existing css in the repo.

epwinchell replied:
@abonas try: = stylesheet_link_tag 'topology'  

where topology.css is in assets/stylesheets

Fryguy commented on app/views/ems_container/_main.html.haml:
```diff
@@ -1 +1,13 @@
+    = render :partial => "shared/summary/textual", :locals => {:title => _("Topology"),
+                                                               :items => textual_group_topology}
+    = render :partial => "shared/summary/textual_tags", :locals => {:title => _("Smart Management"),
+                                                                    :items => textual_group_smart_management}
```
Needs a newline at the end.

abonas replied:
> Needs a newline at the end.

this haml file doesn't "like" a newline in the end, it starts marking the whole file as invalid..

martinpovolny replied:
@abonas : please, recheck, the above statement does not seem true

himdel replied:
Still no newline, I actually checked and haml-lint handles it just fine, what's the problem?

abonas replied:
> Still no newline, I actually checked and haml-lint handles it just fine, what's the problem?

I'm getting to conclusion this is an issue with Rubymine :)

Fryguy commented on app/views/ems_container/show.html.haml:
```diff
@@ -1 +1 @@
-= render :partial => "shared/views/ems_common/show"
+= render :partial => @showtype
```
Also needs newline at the end.

abonas replied:
> Also needs newline at the end.

sure, will add. why the rubocop doesn't alert on it?

abonas commented:
> I really hope the widget can be generalized. I'd expect the legend/list of entities is not the issue and can be dynamic, rather the format on actual entities is currently "kubernetes-like". any reason we can't provide the other entities in a similar format for the widget?

@itamarh 
There are currently 3 aspects to "generalizing" the widget as I see it:
1. (minor) - naming of the component itself that prefixes with "kubernetes"
2. the format of the entities fed to the component - currently it expects a "kubernetes like" json, with a metadata section, etc.
3. potentially using properties/attributes of some entities (for tooltips, extra data, etc) that are not available on kubernetes json data model.

Indeed the legend/entities/entity types list is dynamic, and I already currently "feed" the graph with hosts and VMs which are definitely not part of the kubernetes native entities :) ,  so I build their data in a format of "like kubernetes" that the component expects. So all entities *can* be fed to the graph in a kubernetes like way, it's just a bit of an unnatural way of doing it.

abonas commented:
@chessbyte this is not wip, should have a second round of reviews now

abonas commented:
@h-kataria @Fryguy @epwinchell 
I addressed the review comments and uploaded a newer version, please see if anything else is missing. Generalizing the widget should not be the delaying factor of this PR, it should be discussed separately.
The only issue that I currently see that perhaps requires fixing is that when going directly to topology url it doesn't navigate user to login page, please advise how/if it should be fixed
(going to http://localhost:3000/container_topology directly gives an blank page if not coming after a login page)

abonas commented:
@Fryguy the failing build fails on brakeman with " High       | ems_container/show (EmsContainerController#show) | Dynamic Rend" - is this a new check? what is the issue here/what should be fixed?

Fryguy commented:
Yes, [brakeman](http://brakemanscanner.org/) is a test suite we've just added to do UI static analysis for security issues.

If you run `bundle exec rake test:brakeman` locally you can reproduce the issue.  Or you can run the following might give you nicer output: `brakeman -i .brakeman.ignore -f html -o log/brakeman.html`, which will generate the log/brakeman.html file.

Here is the info on [Dynamic Render Path](http://brakemanscanner.org/docs/warning_types/dynamic_render_paths/).  We've put a number of these in the ignore file for now, and I assume this came from copying from another view/controller?  cc @tenderlove @dclarizio @h-kataria

abonas commented:
> Yes, brakeman is a test suite we've just added to do UI static analysis for security issues.
> 
> If you run bundle exec rake test:brakeman locally you can reproduce the issue. Or you can run the following which gives you nicer output: brakeman -i .brakeman.ignore -f html -o log/brakeman.html, which will generate the log/brakeman.html file.
> 
@Fryguy 
I added my warning to the ignore file - I had to regenerate the whole file from scratch and get the warning from there because of the dynamic fingerprint needed, is there a better way to do it?
Also, the brakeman runs more than 10 minutes - is there anything to speed it up?

abonas commented:
@martinpovolny @h-kataria @Fryguy can someone please explain the travis failures here? It looks like failing on files that weren't touched in this PR, and the same test suite passed several days ago.
I tried to go over the latest PRs to see any related changes but nothing stands out (maybe except for this https://github.com/ManageIQ/manageiq/pull/3254)

martinpovolny commented:
@abonas : could you, please, move the `build_topology` and related into a separate class? It's a piece of logic that probably should not be in a controller. 

I know we have many places with logic in the controller but we are trying to move it out. Just some plain ruby class call it a `presenter` or `service` whatever. 

We are trying to do the same when cleaning up the old code so it would be nice if your could would bring even more quality to our codebase ;-)

Also: It will be much easier to write tests for the code if it's cleanly cut into a separate class.

martinpovolny commented:
@himdel: the test failures seem to be angular (maybe version?) related, can you take a look?

abonas commented:
> could you, please, move the build_topology and related into a separate class? It's a piece of logic that probably should not be in a controller.
> 
> I know we have many places with logic in the controller but we are trying to move it out. Just some plain ruby class call it a presenter or service whatever.
> 
> We are trying to do the same when cleaning up the old code so it would be nice if your could would bring even more quality to our codebase ;-)

@martinpovolny 
Ddo you have a best practices example that should be followed - in the manageiq codebase? 
where should this presenter/service file be placed in the codebase?

martinpovolny commented:
@abonas : see `app/presenters` or `app/services`

abonas commented:
> see app/presenters or app/services

@martinpovolny thanks, the latest patch version has the logic extracted to a service

himdel commented:
@abonas , the reason for the failing tests is the dependency on `rails-assets-kubernetes-topology-graph` - it forces angular to version 1.3.* instead of 1.4.* that we're using. 

You may need to create your own gem or make [them](https://github.com/kubernetes-ui/topology-graph) update the dependency to angular 1.4

abonas commented:
> EDIT: also, I'd like to see the global topologyApp variable gone - please consider chaining the methods instead if you need it only in that file, or puting it under the ManageIQ global object if it's actually needed elsewhere.

It will be great to get an explanation why do you ask for this variable to be gone. 
Part of a code review process is to explain why do you ask for things to be changed.

himdel commented:
@abonas Sure.. we've been doing a lot of Javascript cleanup recently, part of which was to identify all the global variables used in JS and to try to reduce the number. That should be almost done, and all the global state should be in `window.ManageIQ` now. Your PR however introduces a new global variable which goes against that, and since that variable has no reason for existence..

martinpovolny commented:
@abonas : We are trying to cleanup global variables. Removing what we can. Make local what can be local. Group them. Give meaningful names... Also we are trying to get some elementary test coverage for javascript code.

abonas commented:
> @abonas Sure.. we've been doing a lot of Javascript cleanup recently, part of which was to identify all the global variables used in JS and to try to reduce the number. That should be almost done, and all the global state should be in window.ManageIQ now. Your PR however introduces a new global variable which goes against that, and since that variable has no reason for existence..
> 
> EDIT: If you're interested in details, it was #3254.

thanks for explaining. please see the latest version of this patch.

himdel commented:
@abonas great, thanks! :) But.. I just noticed another thing... the `$interval` call - the [docs](https://docs.angularjs.org/api/ng/service/$interval) say that it's not automatically destroyed when the scope gets destroyed, so your refresh will kept getting called immediately.

(You can do `$interval.cancel` with the return value of the original call in `$scope.on('destroy', ...)` handler.)

abonas commented:
> @abonas great, thanks! :) But.. I just noticed another thing... the $interval call - the docs say that it's not automatically destroyed when the scope gets destroyed, so your refresh will keep getting called indefinitely.
> 
> (You can do $interval.cancel with the return value of the original call in $scope.on('destroy', ...) handler.)

thanks for pointing this out, please check latest revision, let me know if that's what you meant

himdel commented on app/assets/javascripts/controllers/container_topology/container_topology_controller.js:
```diff
@@ -0,0 +1,30 @@
+    $interval( $scope.refresh, 1000*60*3);
+
+    $scope.$on('$destroy', function() {
+        $interval.cancel;
```
Almost right .. except `$interval.cancel` is a function - you need to pass it the return value of the original $interval call.. So `var promise = $interval( $scope.refresh, 1000*60*3);` on line 25 and `$interval.cancel(promise)` here.

himdel commented:
@abonas almost :) I added comment to the appropriate line

zeari commented on app/views/layouts/_view_buttons.html.haml:
```diff
@@ -6,7 +6,7 @@
     = render :partial => "layouts/dhtmlxtoolbar", :locals => {:tb => "view_tb", :tb_yaml => "drift_view_tb"}
   - elsif !%w(all_tasks all_ui_tasks timeline diagnostics my_tasks my_ui_tasks miq_server usage).include?(@layout) |
-    && !@layout.starts_with?("miq_request") && !@treesize_buttons |
+    && (@layout.nil? || !@layout.starts_with?("miq_request")) && !@treesize_buttons |
```
@abonas What is the purpose of this change?

zeari commented on app/helpers/application_helper.rb:
```diff
@@ -1002,7 +1002,11 @@ def pdf_page_size_style
                         container_route container_project container_replicator container_image container_image_registry
+=======
+                        container_route container_project container_replicator container_topology
+>>>>>>> Topology widget for containers provider.
```
@abonas This slipped in

zeari commented on app/controllers/application_controller.rb:
```diff
@@ -2201,13 +2201,14 @@ def get_global_session_data
         session[:tab_url][:opt] = inbound_url if ["utilization","planning","bottlenecks","waste"].include?(action_name)
       when "catalog", "vm", "vm_or_template", "miq_template", "service"
-        session[:tab_url][:svc] = inbound_url if ["show", "show_list", "explorer"].include?(action_name)
+        session[:tab_url][:svc] = inbound_url if ["show", "shoshoww_list", "explorer"].include?(action_name)
```
@abonas probably a mistype

abonas commented:
A new upload of topology widget.
Fixed several review comments, moved styling to a separate css, using the new angular from rails-assets.
@epwinchell , @martinpovolny @himdel can you give this another round of review please?
I have already 2 enhancement patches for this in my queue...

himdel commented on Gemfile:
```diff
@@ -108,6 +108,7 @@ source "https://rails-assets.org" do
   gem "rails-assets-bootstrap-datepicker",     "~>1.4.0"
   gem "rails-assets-bootstrap-hover-dropdown", "~>2.0.11"
   gem "rails-assets-bootstrap-select",         "~>1.7.3"
+  gem 'rails-assets-kubernetes-topology-graph', '=0.0.16'
```
In a1d63f2, Brandond made the quoting consistent, could you change to double quotes please?

abonas replied:
> In a1d63f2, Brandon made the quoting consistent, could you change to double quotes please?

sure

himdel commented on app/assets/javascripts/application.js:
```diff
@@ -25,6 +25,9 @@
 //= require controllers/ops/log_collection_form_controller
 //= require controllers/ops/tenant_form_controller
+//= require controllers/container_topology/container_topology_controller
+//= require d3/d3
```
Missing `rails-assets-d3` from Gemfile?

abonas replied:
> Missing rails-assets-d3 from Gemfile?

actually now I noticed that d3/d3 line appears twice so I remove now this one.
As for rails-assets dependency, it supposed to be coming from rails-assets-c3

himdel replied:
Right, missed that, sorry :) `d3` could be enough though, but c3 is as `c3/c3` as well, so ignore me :)

himdel commented on app/assets/javascripts/controllers/container_topology/container_topology_controller.js:
```diff
@@ -0,0 +1,37 @@
+angular.module('topologyApp', ['kubernetesUI'])
+.config(['$httpProvider', function($httpProvider) {
+    $httpProvider.defaults.headers.common['X-CSRF-Token'] = jQuery('meta[name=csrf-token]').attr('content');
```
Hmm, I agree we need this but it feels *wrong* to have to repeat it in every module.. maybe we can do a `miqHttp` service to wrap `$http` or something.. But I guess that's for a separate PR, just thinking out loud here :)

himdel commented on app/helpers/ems_container_helper/textual_summary.rb:
```diff
@@ -75,4 +79,12 @@ def textual_refresh_status
+    h     = {:label => label, :image => "container_topology"}
+    h[:link]  = url_for(:controller => 'container_topology', :action => 'show', :id => @ems.id)
+    h[:title] = "Show topology"
+    h
```
Why not a hash literal?

abonas replied:
> Why not a hash literal?

what do you mean?

himdel replied:
It should probably say:

    def textual_topology
        {:label => N_('Topology'),
         :image => 'container_topology',
         :link  = url_for(:controller => 'container_topology', :action => 'show', :id => @ems.id),
         :title => N_("Show topology")}
    end

abonas replied:
> def textual_topology
>     {:label => N_('Topology'),
>      :image => 'container_topology',
>      :link  = url_for(:controller => 'container_topology', :action => 'show', :id => @ems.id),
>      :title => N_("Show topology")}
> end

sure, I'll change it

himdel commented on app/views/configuration/_ui_2.html.haml:
```diff
@@ -37,8 +37,9 @@
-          - if has_any_role?(%w(ems_container_show_list container_group_show_list container_node_show_list container_route_show_list container_project_show_list container_replicator_show_list container_image_show_list container_image_registry_show_list container_service_accord containers_filter_accord containers_accord))
+          - if has_any_role?(%w(ems_container_show_list container_group_show_list container_node_show_list container_route_show_list container_project_show_list container_replicator_show_list container_image_show_list container_image_registry_show_list container_service_accord containers_filter_accord containers_accord container_topology))
             %fieldset
+<<<<<<< HEAD
```
Bad rebase?

abonas replied:
> Bad rebase?

yap, good catch. surprisingly it didn't crash the UI :)

himdel commented on app/views/configuration/_ui_2.html.haml:
```diff
@@ -50,7 +51,8 @@
                         :containerimageregistry => "container_image_registry",
-                        :container              => "container"}
+                        :container              => "container",
+                        :containertopology   => "container_topology"}
```
Missing whitespace (but I guess I'm just parroting miq-bot now :))

abonas replied:
> Missing whitespace (but I guess I'm just parroting miq-bot now :))

done

martinpovolny commented on app/helpers/application_helper.rb:
```diff
@@ -1214,7 +1215,7 @@ def start_page_allowed?(start_page)
   def allowed_filter_db?(db)
-    return false if db.start_with?('Container') && !get_vmdb_config[:product][:containers]
+    return false if (db.nil? || db.start_with?('Container')) && !get_vmdb_config[:product][:containers]
     true
```
please, rewrite whole method to
```db.to_s.start_with?('Container') || get_vmdb_config[:product][:containers]```

if you are already touching it...

abonas replied:
> please, rewrite whole method to
> !db.to_s.start_with?('Container') || get_vmdb_config[:product][:containers]
> 
> if you are already touching it...

this method is going to be cleaned up anyway in another patch
cc @zakiva @simon3z 
(cleaning leftovers of a patch that was merged too early: https://github.com/ManageIQ/manageiq/pull/3684)
so I prefer not rewrite it as part of topology widget - it's not the primarily focus of this PR and should be treated separately.

martinpovolny replied:
@abonas : Good to know, it's going to get fixed.

However to review your change anyone has to figure out what this method does to understand what your change does.

So generally when dealing with such messy bit either don't touch it at all or clean it up, please.

himdel replied:
This is still broken.. do you actually need the `db.nil?` logic? Feels like you're introducing a bug.. Why should it return false for nil, but true for ""? (Assuming `:product` `:containers` being falsy).

abonas replied:
> This is still broken.. do you actually need the db.nil? logic? Feels like you're introducing a bug.. Why should it return false for nil, but true for ""? (Assuming :product :containers being falsy).

topology is not relying on any rails model/or db table. hence, its db is nil and it is a special use case.

martinpovolny replied:
@abonas : here we go with my initial comment. No **bloody chance** anyone revisiting this code any time later will gues anywhere near this explanation. This should be nicely formatted method with a spec explaining the case or at least a comment..

abonas replied:
> @abonas : here we go with my initial comment. No bloody chance anyone revisiting this code any time later will guess anywhere near this explanation. This should be nicely formatted method with a spec explaining the case or at least a comment..

ok lets try another approach. please have a look at https://github.com/ManageIQ/manageiq/pull/4075
if that will get in, I can rebase and it will resolve the current review issue. how does that sound?

martinpovolny replied:
ok

abonas replied:
> ok

@martinpovolny , https://github.com/ManageIQ/manageiq/pull/4075 got merged and this patch is rebased on it, so this code is gone.

martinpovolny commented on app/presenters/menu/default_menu.rb:
```diff
@@ -90,6 +90,8 @@ def container_menu_section
                          '/container_image_registry'),
+          Menu::Item.new('container_topology', 'Topology', 'container_topology',
+                         {:feature => 'container_topology', :any => true}, '/container_topology')
         ])
```
`N_('Topology')`, please

abonas replied:
> N_('Topology'), please

sure

himdel commented on spec/models/miq_product_feature_spec.rb:
```diff
@@ -5,14 +5,14 @@
       MiqRegion.seed
       MiqProductFeature.seed
-      MiqProductFeature.count.should eq(843)
+      MiqProductFeature.count.should eq(845)
```
If I'm counting right, you're adding just one ;)

abonas commented:
an updated screenshot
![latest_topology](https://cloud.githubusercontent.com/assets/6277245/9495296/9c495c7c-4c13-11e5-9c5b-2586f9fe6586.png)

abonas commented:
I think I addressed all review comments here.
Please let me know what else is missing to get this merged.
The coveralls for some reason drops every run even more even when no new code is introduced. it started at -0.0002, then proceeded to -7 and now dropped to -14, so it's clearly not functioning properly.

himdel commented on app/assets/javascripts/application.js:
```diff
@@ -47,7 +50,6 @@
 //= require bootstrap-select
 //= require bootstrap-hover-dropdown
 //= require c3/c3
-//= require d3/d3
```
Sorry, could we keep c3 and d3 together? And you still have single quotes in the Gemfile, albeit only for version now.

abonas replied:
> Sorry, could we keep c3 and d3 together? And you still have single quotes in the Gemfile, albeit only for version now.

will do. will take c3 up to d3. apparently the order in this file matters. having d3 down next to c3 causes the topology not to be drawn..

matthewd commented on app/helpers/ems_container_helper/textual_summary.rb:
```diff
@@ -27,6 +27,10 @@ def textual_group_smart_management
+  def textual_group_topology
+    items = %w(topology)
+    items.collect { |m| send("textual_#{m}") }.flatten.compact
+  end
```
This probably won't actually conflict, but please rebase per #4036

abonas replied:
> This probably won't actually conflict, but please rebase per #4036

of course, will rebase

abonas commented:
rebased, fixed more rucobop warnings, fixed feature spec count.

abonas commented:
cc @himdel @martinpovolny

abonas commented:
@martinpovolny @himdel @epwinchell @dclarizio please let me know what else is needed in this PR so it can be merged, I have 2 subsequent enhancement PRs in my queue for topology that wait for this one.
Thank you

martinpovolny commented:
I have no further comments. Thx

abonas commented:
> I have no further comments. Thx

wonderful :balloon: 
who can merge it?

martinpovolny commented:
@abonas : I can. But travis is red:
```
1) MiqProductFeature.seed empty table
     Failure/Error: MiqProductFeature.count.should eq(848)
       
       expected: 848
            got: 849
       
       (compared using ==)
     # ./spec/models/miq_product_feature_spec.rb:8:in `block (3 levels) in <top (required)>'
  2) MiqProductFeature.seed run twice
     Failure/Error: MiqProductFeature.count.should eq(848)
       
       expected: 848
            got: 849
       
       (compared using ==)
     # ./spec/models/miq_product_feature_spec.rb:15:in `block (3 levels) in <top (required)>'
  3) MiqProductFeature.seed with existing records
     Failure/Error: expect { deleted.reload }.to raise_error(ActiveRecord::RecordNotFound)
     NoMethodError:
       undefined method `expect' for #<RSpec::Matchers::BuiltIn::Eq:0x0000002f0f19a8 @expected=848>
     # ./spec/models/miq_product_feature_spec.rb:27:in `block (3 levels) in <top (required)>'
```

abonas commented:
> @abonas : I can. But travis is red:

sorry, missed that. too many rebases to the same patch... will fix in a sec.

martinpovolny commented on app/assets/stylesheets/container_topology.css:
```diff
@@ -0,0 +1,48 @@
+    width: 39px !important;;
+    height: 39px !important;;
+    display: block;
+}
```
missing end-of-line

abonas replied:
> missing end-of-line

added

martinpovolny commented:
@abonas : Sorry, I have one more issue with this PR. There are no tests at all. You should at least:

  1. provide a test to check that `app/views/container_topology/show.html.haml` actually renders. Send a get/post/whatever is needed request to the controller and test that the view is rendered w/o error and contains the very basis stuff that you expect (correct view is rendered etc).
  1. provide unit tests for `app/services/container_topology_service.rb`

Once you do that, I can merge your PR. I will be on-line for at least 6 hours today and will respond to your changes, questions, whatever.

abonas commented:
> provide a test to check that app/views/container_topology/show.html.haml actually renders. Send a get/post/whatever is needed request to the controller and test that the view is rendered w/o error and contains the very basis stuff that you expect (correct view is rendered etc).

I just uploaded a test for that part.
I'll work on the tests for service now.

chessbyte commented:
Have not looked into details, but want to insure that the direction of this (and future PRs) is to provide a reusable topology widget that we can use not only in Containers, but also in Cloud, Virtual Infra, Storage and Network.  /cc @abonas @martinpovolny @Fryguy @matthewd

Just checking in from vacation in Caribbean.  :-)

abonas commented:
> Have not looked into details, but want to insure that the direction of this (and future PRs) is to provide a reusable topology widget that we can use not only in Containers, but also in Cloud, Virtual Infra, Storage and Network. /cc @abonas @martinpovolny @Fryguy @matthewd
> 
> Just checking in from vacation in Caribbean. :-)

I am reminding that this PR just uses the topology widget code that is external to manageiq.
This PR uses topology widget ver 0.0.16, and it already does reuse the widget generically (it adds hosts and vms to its view)
topology widget version 0.0.17 allows complete generic reusability, and it will be introduced to ManageIQ in 2 PRs after this one.

abonas commented:
> @abonas : Sorry, I have one more issue with this PR. There are no tests at all. You should at least:
> 
> provide a test to check that app/views/container_topology/show.html.haml actually renders. Send a get/post/whatever is needed request to the controller and test that the view is rendered w/o error and contains the very basis stuff that you expect (correct view is rendered etc).
> provide unit tests for app/services/container_topology_service.rb
> Once you do that, I can merge your PR. I will be on-line for at least 6 hours today and will respond to your changes, questions, whatever.

@martinpovolny 
I added all tests as requested. Please have a look.

abonas commented on spec/services/container_topology_service_spec.rb:
```diff
@@ -0,0 +1,60 @@
+    end
+  end
+
+end
```
> Looks good except for the missing end-of-line.
> 
> Can you, please, check this link? https://robots.thoughtbot.com/no-newline-at-end-of-file
> 
> There seems to be an advice to setup RubyMine. You use RubyMine, right?

will do. yes I use Rubymine.

abonas commented:
@martinpovolny  everything is ready

martinpovolny commented:
@abonas : I went through the tests and other bits once again and it looks good. But travis is still failing. Can you look into it, please?

abonas commented:
> @abonas : I went through the tests and other bits once again and it looks good. But travis is still failing. Can you look into it, please?

I repushed again with an updated feature count. it really is an annoying check.

martinpovolny commented:
@abonas: can you, please, take a look at the Rubocop output above? There seem to be several relevant points.

abonas commented:
> @abonas: can you, please, take a look at the Rubocop output above? There seem to be several relevant points.

I tried to fix a few, the problem with those, you fix one, it creates another one, and then need to wait again for the tests to complete, repush, etc. it is an endless story.
I would really like to stop spending time on this patch.

abonas commented:
@miq-bot add_label topology
